File: strings/medium/golang_solutions/sliding_window/palindromic_substrings_647.go
package sliding_window

func countSubstrings(s string) (sum int) {
	if len(s) == 0 {
		return
	}
	sum += len(s)

	windowSize := 2

	for windowSize <= len(s) {
		for i := 0; i+windowSize <= len(s); i++ {
			if isPalindromic(s[i : i+windowSize]) {
				sum++
			}
		}
		windowSize++
	}
	return
}

func countSubstringsBetter(s string) (sum int) {
	if len(s) == 0 {
		return 0
	}
	for i := 0; i < len(s); i++ {
		extendPalindrome(s, i, i, &sum)
		extendPalindrome(s, i, i+1, &sum)
	}
	return sum
}

func extendPalindrome(str string, left int, right int, sum *int) {
	for left >= 0 && right < len(str) && str[left] == str[right] {
		*sum++
		left--
		right++
	}
}

func isPalindromic(str string) bool {
	left, right := 0, len(str)-1

	for left <= right {
		if str[left] != str[right] {
			return false
		}
		left++
		right--
	}
	return true
}

File: leetcode/strings/medium/java_solutions/PalindromicSubstrings_647.java
package leetcode.strings.medium.java_solutions;

public class PalindromicSubstrings_647 {

    private int sum;

    public int countSubstrings(String s) {
        int N = s.length();

        for (int i = 0; i < N; i++) {
            extendSubstring(s, i, i);
            extendSubstring(s, i, i + 1);
        }
        return sum;
    }

    private void extendSubstring(String s, int left, int right) {
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            sum++;
            left--;
            right++;
        }
    }

    // ----------------------------------------------------------------------------------------

    public int countSubstringsBruteForce(String s) {
        int window = 1, result = 0;

        while (window <= s.length()) {
            for (int i = 0; i + window <= s.length(); i++) {
                String substring = s.substring(i, i + window);
                if (isPalindromic(substring)) {
                    result++;
                }
            }
        }
        return result;
    }

    private boolean isPalindromic(String substring) {
        int left = 0, right = substring.length() - 1;

        while (left < right) {
            if (substring.charAt(left) != substring.charAt(right)) return false;
            left++;
            right--;
        }
        return true;
    }
}
